### PR TITLE
Feat/lower express threshold

### PR DESCRIPTION
**Pull request by karlnyr**
## Description

The start of express cases should not be haltered unless we suspect that the data gained is corrupt and inconsistent. To adress this, we lower the read count threshold to 50% of the target number of reads (according to apptag) , and if the number of reads is below this threshold then an investigation is required to valide a start of the case.

### Changed

- Read count threshold to start Express-cases automatically to 50% 

### How to prepare for test

- [x] ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] paxa the environment: `paxa`
- [x] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh feat/lower-express-threshold`

### How to test

- [x] Set the number of reads of an "non-express"-sample to: 75% expected > reads > 50% expected reads.
- [x] Set the case status to "analyze"
- [x] run `cg workflow <relevant pipeline> start-available`
- [x] Set the priority to express for the sample
- [x] Set the case status to "analyze"
- [x] run `cg workflow <relevant pipeline> start-available`
- [x] Set the number of reads of the express sample to: 50% expected reads > reads

### Expected test outcome

- [x] "Non-Express"-Case is not started when reads are:  75% expected > reads > 50% expected reads
- [x] Express-case is not started when reads are: 50% expected reads > reads
- [x] Express-case is started when reads are: 75% expected > reads > 50% expected reads 

## Review

- [x] code approved by HS, MR
- [x] tests executed by @karlnyr 
- [x] "Merge and deploy" approved by HS, MR

Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in [production/data-analysis/workflow/mip](https://atlas.scilifelab.se/production/data_analysis/workflows/MIP_analysis/), comment in BALSAMIC AM doc `2046`
- [ ] Deploy this branch
- [ ] Inform to @Clinical-Genomics/data-analysis

### CODE REVIEW

**Comment by karlnyr** [comment]
Case for `ACC8944A2` is `procondor`. Conditions are set for `cg workflow mip-dna start-available` to pick up the case by setting the status to `Analyze` on the case. The one thing that should stop the case is read counts which is set to be:  75% expected > read count > 50% expected
  
image

Case was not picked up. 

Changing priority of the sample to express, setting the read count: 50% expected > read count:
image

Setting the new threshold read count: 75% expected > read count > 50% expected:

image

Case was picked up!

File: cg/store/models.py [hunk 1]
```diff
@@ -509,8 +509,13 @@ def __str__(self) -> str:
...
         """Return sequencing qc passed or failed."""
-        application = self.application_version.application
-        return self.reads > application.expected_reads
+        if self.priority < PRIORITY_MAP["express"]:
+            application = self.application_version.application
+            return self.reads > application.expected_reads
+        else:
+            application = self.application_version.application
+            fifty_percent = application.target_reads / 2
+            return self.reads > fifty_percent
```
Comment (henrikstranneheim) [inline cg/store/models.py:518]:
```suggestion
            one_half_target_reads = application.target_reads / 2
            return self.reads > one_half_target_reads
```

**Review by henrikstranneheim — COMMENTED** [review]
Great! Would be nice with a unit test to cover the new logic

File: cg/store/models.py [hunk 2]
```diff
@@ -509,8 +509,13 @@ def __str__(self) -> str:
     @property
     def sequencing_qc(self) -> bool:
         """Return sequencing qc passed or failed."""
-        application = self.application_version.application
```
[the remark this answers is not in the document]
Comment (karlnyr) [reply cg/store/models.py:512]: agreed

File: cg/store/models.py [hunk 3]
```diff
@@ -509,8 +509,13 @@ def __str__(self) -> str:
...
-        return self.reads > application.expected_reads
+        if self.priority < PRIORITY_MAP["express"]:
+            application = self.application_version.application
+            return self.reads > application.expected_reads
+        else:
+            application = self.application_version.application
```
[the remark this answers is not in the document]
Comment (karlnyr) [reply cg/store/models.py]: fixed smell

**Comment by karlnyr** [comment]
> Great! Would be nice with a unit test to cover the new logic

Agreed, will do that asap!

**Comment by karlnyr** [comment]
@henrikstranneheim I believe that the tests added are covering the new business logic 👍

**Review by henrikstranneheim — COMMENTED** [review]
👍 . Just some minor stuff that needs fixing. Good work!

File: tests/store/api/test_store_api_status_analysis.py [hunk 4]
```diff
@@ -172,6 +173,44 @@ def test_one_of_one_sequenced_samples(base_store: Store, helpers):
...
+    test_case = helpers.add_case(base_store)
+    test_sample = helpers.add_sample(base_store, sequenced_at=datetime.now())
+    base_store.relate_sample(test_case, test_sample, "unknown")
+    application = test_sample.application_version.application
+    application.target_reads = 40
+    test_sample.reads = 20
+    test_sample.priority = PRIORITY_MAP["express"]
+
+    # WHEN getting cases to analyse
+    sequencing_qc_ok: bool = test_sample.sequencing_qc
```
[the remark this answers is not in the document]
Comment (karlnyr) [reply tests/store/api/test_store_api_status_analysis.py:189]: This is a property, so I will leave it as is

File: tests/store/api/test_store_api_status_analysis.py [hunk 5]
```diff
@@ -172,6 +173,44 @@ def test_one_of_one_sequenced_samples(base_store: Store, helpers):
...
+    test_case = helpers.add_case(base_store)
+    test_sample = helpers.add_sample(base_store, sequenced_at=datetime.now())
+    base_store.relate_sample(test_case, test_sample, "unknown")
+    application = test_sample.application_version.application
+    application.target_reads = 40
+    test_sample.reads = 20
+    test_sample.priority = PRIORITY_MAP["standard"]
+
+    # WHEN getting cases to analyse
+    sequencing_qc_ok: bool = test_sample.sequencing_qc
```
[the remark this answers is not in the document]
Comment (karlnyr) [reply tests/store/api/test_store_api_status_analysis.py:208]: This is a property, so I will leave it as is

File: tests/store/api/test_store_api_status_analysis.py [hunk 6]
```diff
@@ -172,6 +173,44 @@ def test_one_of_one_sequenced_samples(base_store: Store, helpers):
...
+    assert sequencing_qc_ok
+
+
+def test_normal_case_with_low_reads(base_store: Store, helpers):
+    """Test that express cases with reads more than 50% of the target reads are among the cases to analyse"""
+
+    # GIVEN a database with a case which has an express sample with half the amount of reads
+    test_case = helpers.add_case(base_store)
+    test_sample = helpers.add_sample(base_store, sequenced_at=datetime.now())
+    base_store.relate_sample(test_case, test_sample, "unknown")
```
[the remark this answers is not in the document]
Comment (karlnyr) [reply tests/store/api/test_store_api_status_analysis.py:201]: That refers to the status of the the status of the sample, i.e `affected`, `unaffected`. In the new tests I have done I removed the cases all together, as they were unnecessary for the purpose of looking at a samples QC.

**Review by Mropat — APPROVED** [review]

**Review by henrikstranneheim — APPROVED** [review]
👍

**Comment by karlnyr** [comment]
Installed on prod:
image

**Pull request merged**